homeassistant/components/diagnostics/util.py

```diff
@@ -2,27 +2,34 @@
 from __future__ import annotations
 
 from collections.abc import Iterable, Mapping
-from typing import Any
+from copy import deepcopy
+from typing import Any, TypeVar
 
 from homeassistant.core import callback
 
 from .const import REDACTED
 
+T = TypeVar("T")
+
 
 @callback
-def async_redact_data(data: Mapping, to_redact: Iterable[Any]) -> dict[str, Any]:
+def async_redact_data(data: T, to_redact: Iterable[Any]) -> T:
     """Redact sensitive data in a dict."""
     if not isinstance(data, (Mapping, list)):
         return data
 
-    redacted = {**data}
-
-    for key, value in redacted.items():
-        if key in to_redact:
-            redacted[key] = REDACTED
-        elif isinstance(value, dict):
-            redacted[key] = async_redact_data(value, to_redact)
-        elif isinstance(value, list):
-            redacted[key] = [async_redact_data(item, to_redact) for item in value]
+    redacted = deepcopy(data)
```

Comment:
We should not deepcopy it. That's making a lot of copies. Instead, inside the `isinstance`, you know what it is. Make a copy there.